### PR TITLE
[arrow_dataset.py] add new filter function

### PR DESCRIPTION
**Pull request by patrickvonplaten**
The `.map()` function is super useful, but can IMO a bit tedious when filtering certain examples.
I think, filtering out examples is also a very common operation people would like to perform on datasets.

This PR is a proposal to add a `.filter()` function in the same spirit than the `.map()` function.

Here is a sample code you can play around with:

```python
ds = nlp.load_dataset("squad", split="validation[:10%]")


def remove_under_idx_5(example, idx):
    return idx < 5


def only_keep_examples_with_is_in_context(example):
    return "is" in example["context"]


result_keep_only_first_5 = ds.filter(remove_under_idx_5, with_indices=True, load_from_cache_file=False)
result_keep_examples_with_is_in_context = ds.filter(only_keep_examples_with_is_in_context, load_from_cache_file=False)

print("Original number of examples: {}".format(len(ds)))
print("First five examples number of examples: {}".format(len(result_keep_only_first_5)))
print("Is in context examples number of examples: {}".format(len(result_keep_examples_with_is_in_context)))
```

### CODE REVIEW

File: src/nlp/arrow_dataset.py [hunk 1]
```diff
@@ -557,3 +558,43 @@ def apply_function_on_filtered_inputs(inputs, indices):
...
+
+    def filter(
+        self,
+        function,
+        with_indices=False,
+        **kwargs
+    ):
+        def map_function(batch, *args):
+            # flatten dict to list
+            flattened_batch = flatten_nested(batch)
```
Comment (lhoestq) [inline src/nlp/arrow_dataset.py:570]: I'm not a big fan of using `flatten_nested` here

File: src/nlp/arrow_dataset.py [hunk 2]
```diff
@@ -557,3 +558,43 @@ def apply_function_on_filtered_inputs(inputs, indices):
...
+            num_items = len(batch.keys())
+            num_examples = len(flattened_batch) // num_items
+            result = defaultdict(list)
+
+            # create single examples
+            for example_pos in range(num_examples):
+                example = {}
+                for i, key in enumerate(batch.keys()):
+                    example[key] = flattened_batch[i * num_examples + example_pos]
```
Comment (lhoestq) [inline src/nlp/arrow_dataset.py:579]: It's probably not going to work for nested structures

**Comment by lhoestq** [comment]
I agree that a `.filter` method would be VERY useful and appreciated. I'm not a big fan of using `flatten_nested` as it completely breaks down the structure of the example and it may create bugs. Right now I think it may not work for nested structures. Maybe there's a simpler way that we've not figured out yet.

**Comment by lhoestq** [comment]
Instead of flattening everything and rebuilding the example, maybe we can try to access the examples like this:
```python
for i in range(num_examples):
    example = map_nested(lambda x: x[i], batch)
    # ... then test to keep it or not
```

File: src/nlp/arrow_dataset.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/nlp/arrow_dataset.py:579]: Works for `trivia_qa` as well :-) It essentially just merges the value lists of all values in a `dict`

**Comment by patrickvonplaten** [comment]
> Instead of flattening everything and rebuilding the example, maybe we can try to access the examples like this:
> 
> ```python
> for i in range(num_examples):
>     example = map_nested(lambda x: x[i], batch)
>     # ... then test to keep it or not
> ```

Awesome I'll check it out :-)

File: src/nlp/arrow_dataset.py (hunk 2 above)
Comment (lhoestq) [reply src/nlp/arrow_dataset.py:579]: Ok I see but if we have a batch of size two like this one:
```python
{
   "qa": {
        "question": ["foo", "bar],
        "answer": ["1", "2"]
    }
}
```
Then the reconstructed examples will be different as `flatten_nested` will return something like `["foo", "bar", "1", "2"]` and you only reconstruct the first depth of the structure (`len(batch.keys())`=1)

**Comment by patrickvonplaten** [comment]
> Instead of flattening everything and rebuilding the example, maybe we can try to access the examples like this:
> 
> ```python
> for i in range(num_examples):
>     example = map_nested(lambda x: x[i], batch)
>     # ... then test to keep it or not
> ```

Awesome this function is definitely much nicer!

**Comment by lhoestq** [comment]
Actually I just realized that `map_nested` might not work either as it applies the function at the very las list of the structure. However we can imagine that a single example has also a list in its structure:
```python
one_example = {
    "title": "blabla",
    "paragraphs": [
        "p1", "p2", ...
    ]
}
```

**Comment by lhoestq** [comment]
We'll probably have to take into account the `dset._data.schema` to extract the examples from the batch.

**Comment by patrickvonplaten** [comment]
> Actually I just realized that `map_nested` might not work either as it applies the function at the very last list of the structure. However we can imagine that a single example has also a list in its structure:
> 
> ```python
> one_example = {
>     "title": "blabla",
>     "paragraphs": [
>         "p1", "p2", ...
>     ]
> }
> ```

They both work. I'm using it on trivia_qa which is pretty nested. If you use the option `with_dict=True` I think it's fine.

File: src/nlp/arrow_dataset.py [hunk 3]
```diff
@@ -557,3 +558,39 @@ def apply_function_on_filtered_inputs(inputs, indices):
...
             return self
+
+    def filter(self, function, with_indices=False, **kwargs):
+        def map_function(batch, *args):
+            result = defaultdict(list)
+            num_examples = len(batch[next(iter(batch.keys()))])
+
+            # create single examples
+            for i in range(num_examples):
+                example = map_nested(lambda x: x[i], batch, dict_only=True)
```
Comment (patrickvonplaten) [inline src/nlp/arrow_dataset.py:569]: So we assume that `batch` is a `dict{key_1: list, key_2: list, ...}`. With `dict_only=True` this means that the function has to be applied directly to every list in the `dict` and it should be fine @lhoestq

**Comment by patrickvonplaten** [comment]
> We'll probably have to take into account the `dset._data.schema` to extract the examples from the batch.

Why?

File: src/nlp/arrow_dataset.py (hunk 3 above)
Comment (lhoestq) [reply src/nlp/arrow_dataset.py:569]: Ok I see what you mean !

**Comment by lhoestq** [comment]
Actually it's fine. I guess this is going to be yet another thing to be unit-tested just to make sure ^^

**Comment by patrickvonplaten** [comment]
Yes, I will need to add tests and documentation! 
@thomwolf - would a function like this be ok? It abstracts `.map()` a bit which might be hard to understand.

**Comment by lhoestq** [comment]
I tried on some datasets with nested structure and it works fine ! Great work :D

**Comment by patrickvonplaten** [comment]
Awesome :-), I will add documentation and some simple unittests

File: src/nlp/arrow_dataset.py [hunk 4]
```diff
@@ -557,3 +558,66 @@ def apply_function_on_filtered_inputs(inputs, indices):
                 return Dataset.from_buffer(buf_writer.getvalue())
         else:
             return self
+
+    def filter(self, function, with_indices=False, **kwargs):
+        """ Apply a filter function to all the elements in the table in batches
+            and update the table so that the dataset only includes examples according to the filter function.
+
+            Args:
```
Comment (patrickvonplaten) [inline src/nlp/arrow_dataset.py:566]: added docstring here

File: tests/test_arrow_dataset.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+import numpy as np
+import pyarrow as pa
+
+from nlp.arrow_reader import BaseReader
+from nlp.info import DatasetInfo
+from nlp.splits import SplitDict, SplitInfo
+
+
+class ReaderTester(BaseReader):
```
Comment (patrickvonplaten) [inline tests/test_arrow_dataset.py:13]: @lhoestq - it's slightly different from what you used so I copy pasted the code. I guess we could refactor this into `test_utils.py` at one point.

File: tests/test_arrow_dataset.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            filename_skip_take["skip"] if "skip" in filename_skip_take else None,
+            filename_skip_take["take"] if "take" in filename_skip_take else None,
+        )
+        pa_table = pa.Table.from_pydict({"filename": [filename + "_" + str(x) for x in np.arange(100).tolist()]})
+        if skip is not None and take is not None:
+            pa_table = pa_table.slice(skip, take)
+        return pa_table
+
+
+class BaseDatasetTest(TestCase):
```
Comment (patrickvonplaten) [inline tests/test_arrow_dataset.py:32]: @thomwolf @lhoestq  - added 3 tests for `.map()` for now and 1 test for `.filter()`

**Comment by patrickvonplaten** [comment]
Ok merging!

File: datasets/wikipedia/wikipedia.py [hunk 7]
```diff
@@ -25,9 +25,9 @@
 import xml.etree.cElementTree as etree
 
 import apache_beam as beam
-import mwparserfromhell
 import six
 
+import mwparserfromhell
```
Comment (lhoestq) [inline datasets/wikipedia/wikipedia.py:30]: hmmmmmmmmmmmmmmm

**Pull request merged**